runtime/common/src/migrations/hold_reason.rs

```diff
@@ -0,0 +1,151 @@
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use frame_support::pallet_prelude::ValueQuery;
+use frame_support::{storage_alias, Blake2_128Concat, BoundedVec, Parameter};
+
+use cfg_primitives::{AccountId, Balance};
+use frame_support::traits::{ConstU32, Get, Len, OnRuntimeUpgrade, VariantCountOf};
+use pallet_balances::IdAmount;
+use pallet_order_book::weights::Weight;
+use pallet_transfer_allowlist::HoldReason;
+#[cfg(feature = "try-runtime")]
+use parity_scale_codec::{Decode, Encode};
+use parity_scale_codec::{FullCodec, FullEncode};
+use sp_runtime::traits::Member;
+use sp_runtime::SaturatedConversion;
+#[cfg(feature = "try-runtime")]
+use sp_runtime::Saturating;
+#[cfg(feature = "try-runtime")]
+use sp_runtime::TryRuntimeError;
+use sp_std::vec;
+use sp_std::vec::Vec;
+
+const LOG_PREFIX: &str = "MigrateTransferAllowList HoldReason:";
+
+pub struct MigrateTransferAllowListHolds<T, RuntimeHoldReason>(
+	sp_std::marker::PhantomData<(T, RuntimeHoldReason)>,
+);
+
+type OldHolds = BoundedVec<IdAmount<(), Balance>, ConstU32<10>>;
+type NewHolds<RuntimeHoldReason> =
+	BoundedVec<IdAmount<RuntimeHoldReason, Balance>, VariantCountOf<RuntimeHoldReason>>;
+
+#[storage_alias]
+pub type Holds<T: pallet_balances::Config> =
+	StorageMap<pallet_balances::Pallet<T>, Blake2_128Concat, AccountId, OldHolds, ValueQuery>;
+
+impl<T, RuntimeHoldReason> OnRuntimeUpgrade for MigrateTransferAllowListHolds<T, RuntimeHoldReason>
+where
+	T: pallet_balances::Config<Balance = Balance, RuntimeHoldReason = RuntimeHoldReason>
+		+ pallet_transfer_allowlist::Config
+		+ frame_system::Config<AccountId = AccountId>,
+	<T as pallet_balances::Config>::RuntimeHoldReason: From<pallet_transfer_allowlist::HoldReason>,
+	RuntimeHoldReason: frame_support::traits::VariantCount
+		+ FullCodec
+		+ FullEncode
+		+ Parameter
+		+ Member
+		+ sp_std::fmt::Debug,
+{
+	fn on_runtime_upgrade() -> Weight {
+		let transfer_allowlist_accounts: Vec<AccountId> =
+			pallet_transfer_allowlist::AccountCurrencyTransferAllowance::<T>::iter_keys()
+				.map(|(a, _, _)| a)
+				.collect();
+		let mut weight =
+			T::DbWeight::get().reads(transfer_allowlist_accounts.len().saturated_into());
+
+		pallet_balances::Holds::<T>::translate::<OldHolds, _>(|who, holds| {
+			if Self::account_can_be_migrated(&who, &transfer_allowlist_accounts) {
+				log::info!("{LOG_PREFIX} Migrating hold for account {who:?}");
+				let id_amount = IdAmount::<RuntimeHoldReason, Balance> {
+					id: HoldReason::TransferAllowance.into(),
+					// Non-emptiness ensured above
+					amount: holds[0].amount,
+				};
+				weight.saturating_accrue(T::DbWeight::get().reads_writes(1, 1));
+
+				Some(NewHolds::truncate_from(vec![id_amount]))
+			} else {
```

Comment:
`if Self::account_can_be_migrated(&who, &transfer_allowlist_accounts)` is false, then if the old Hold storage still has an entry, that means that we need to migrate it with `pallet_restricted_tokens::HoldReason::NativeIndex.into()` because it means that someone used that pallet to hold something thought that pallet.

Or am I wrong in my assumption?

The flow will be:
- if hold in balances and account in allowlist
  - then: `pallet_transfer_allowlist::HoldReason::TransferAllowance.into()`
- else if hold in balances
  - then: `pallet_restricted_tokens::HoldReason::NativeIndex.into()`
else 
  - then: `None`
  
Still scared about how can we know that an account that has been used by transfer allowance has not also be used by restricted_tokens 🤔

Reply:
> `if Self::account_can_be_migrated(&who, &transfer_allowlist_accounts)` is false, then if the old Hold storage still has an entry, that means that we need to migrate it with `pallet_restricted_tokens::HoldReason::NativeIndex.into()` because it means that someone used that pallet to hold something thought that pallet.
> 
> Or am I wrong in my assumption?

We have three code paths which use `Currency::hold` in our codebase:
1. `pallet-rewards` when [depositing stake](https://github.com/centrifuge/centrifuge-chain/blob/94ee4d7fa555bb2504d10bcea22e002dac948926/pallets/rewards/src/lib.rs#L314)
2. `pallet-transfer-allowlist` when [adding a transfer filter](https://github.com/centrifuge/centrifuge-chain/blob/5ccb3d05378d88fbd7e2876665b39651268069c5/pallets/transfer-allowlist/src/lib.rs#L359)
3. `pallet-orderbook` when [placing](https://github.com/centrifuge/centrifuge-chain/blob/5ccb3d05378d88fbd7e2876665b39651268069c5/pallets/order-book/src/lib.rs#L441) or [updating an order](https://github.com/centrifuge/centrifuge-chain/blob/5ccb3d05378d88fbd7e2876665b39651268069c5/pallets/order-book/src/lib.rs#L485).

All three codepaths use the `RestrictedTokens` as their `Currency` trait which provides two implementations for `hold`: One for [native currency](https://github.com/centrifuge/centrifuge-chain/blob/f713417d8a25b40fe54779ca775f5ae723b8ad73/pallets/restricted-tokens/src/impl_fungible.rs#L280) (i.e. fungible which falls back to the impl of `pallet-balances`) and one for [non-native assets](https://github.com/centrifuge/centrifuge-chain/blob/f713417d8a25b40fe54779ca775f5ae723b8ad73/pallets/restricted-tokens/src/impl_fungibles.rs#L446) (i.e. fungibles which falls back to the impl of `orml-tokens`). The crucial part is that the `orml-tokens` implementation of `hold` [has not made use of the `RuntimeHoldReason`](https://github.com/open-web3-stack/open-runtime-module-library/blob/f2fc5ee47a228bfe899904f88077f602bcfdca07/tokens/src/lib.rs#L2023) (it does not even have storage for that) such that we only need to migrate the `Holds` storage of `pallet-balances`. 

> The flow will be:
> 
> * if hold in balances and account in allowlist
>   * then: `pallet_transfer_allowlist::HoldReason::TransferAllowance.into()`
> * else if hold in balances
>   * then: `pallet_restricted_tokens::HoldReason::NativeIndex.into()`
> * else
>   * then: `None`

Now coming back to the three codepaths which make use of `hold` API: Both `pallet-rewards` and `pallet-orderbook` only hold fungibles because staking currency is not native and we can assume no `CFG <-> *` orders to exist until we execute the RU. So there can only be a single reason why there exists an entry in `pallet_balances::Holds` storage which points back to the transfer allowlist. There cannot be any other stored holds.

Reply:
Thanks for your reasoning!

Just one thing. Root can call the extrinsic `restricted_tokens::set_balance()` (not sure if this already happened), modifying the hold amount in the account for CFG

Reply:
I guess we never use `orderbook` to create an order with CFG